File: shared/main/java/net/ccbluex/liquidbounce/ui/client/altmanager/sub/GuiBannedServers.java
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.ui.client.altmanager.sub;

import java.awt.*;
import java.io.IOException;

import net.ccbluex.liquidbounce.LiquidBounce;
import net.ccbluex.liquidbounce.api.minecraft.client.gui.IGuiButton;
import net.ccbluex.liquidbounce.api.minecraft.client.gui.IGuiTextField;
import net.ccbluex.liquidbounce.api.util.WrappedGuiScreen;
import net.ccbluex.liquidbounce.api.util.WrappedGuiSlot;
import net.ccbluex.liquidbounce.ui.client.altmanager.GuiAltManager;
import net.ccbluex.liquidbounce.ui.font.Fonts;
import net.ccbluex.liquidbounce.utils.login.MinecraftAccount;
import net.ccbluex.liquidbounce.utils.render.RenderUtils;

import org.lwjgl.input.Keyboard;

public class GuiBannedServers extends WrappedGuiScreen
{
	public String status = "\u00A77Idle...";
	private final GuiAltManager prevGui;
	private final MinecraftAccount account;
	GuiServersList serversList;

	public GuiBannedServers(final GuiAltManager gui, final MinecraftAccount acc)
	{
		prevGui = gui;
		account = acc;
	}

	@Override
	public final void initGui()
	{
		final int width = getRepresentedScreen().getWidth();
		final int height = getRepresentedScreen().getHeight();
		serversList = new GuiServersList(this, account);
		serversList.represented.registerScrollButtons(7, 8);

		final int j = 22;
		getRepresentedScreen().getButtonList().add(classProvider.createGuiButton(1, width - 80, 46, 70, 20, "Add"));
		getRepresentedScreen().getButtonList().add(classProvider.createGuiButton(2, width - 80, 70, 70, 20, "Remove"));

		getRepresentedScreen().getButtonList().add(classProvider.createGuiButton(0, width - 80, height - 65, 70, 20, "Back"));
	}

	@Override
	public final void drawScreen(final int mouseX, final int mouseY, final float partialTicks)
	{
		final int width = representedScreen.getWidth();
		representedScreen.drawBackground(0);

		serversList.represented.drawScreen(mouseX, mouseY, partialTicks);

		Fonts.font40.drawCenteredString("\u00A7cBanned servers\u00A78 of \u00A7a" + (account.getAccountName() == null ? account.getName() : account.getAccountName()), width / 2, 6, 0xffffff);
		Fonts.font35.drawCenteredString("\u00A7a" + (account.getAccountName() == null ? account.getName() : account.getAccountName()) + "\u00A78 is \u00A7cbanned\u00A78 from \u00A7c" + serversList.getSize() + "\u00A7a servers.", width / 2, 18, 0xffffff);
		Fonts.font35.drawCenteredString(status, representedScreen.getWidth() / 2, 32, 0xffffff);

		super.drawScreen(mouseX, mouseY, partialTicks);
	}

	@Override
	public final void actionPerformed(final IGuiButton button) throws IOException
	{
		switch (button.getId())
		{
			case 0:
				mc.displayGuiScreen(prevGui.representedScreen);
				break;
			case 1:
				mc.displayGuiScreen(new GuiAddBanned(this, account).representedScreen);
				break;
			case 2:
				if (serversList.getSelectedSlot() != -1 && serversList.getSelectedSlot() < serversList.getSize())
				{
					account.getBannedServers().remove(account.getBannedServers().get(serversList.getSelectedSlot()));
					LiquidBounce.fileManager.saveConfig(LiquidBounce.fileManager.accountsConfig);
					status = "\u00A7aThe server has been removed.";
				}
				else
					status = "\u00A7cSelect a server.";
				break;
		}
		super.actionPerformed(button);
	}

	@Override
	public final void keyTyped(final char typedChar, final int keyCode) throws IOException
	{
		switch (keyCode)
		{
			case Keyboard.KEY_ESCAPE:
				mc.displayGuiScreen(prevGui.representedScreen);
				return;
			case Keyboard.KEY_UP:
			{
				int i = serversList.getSelectedSlot() - 1;
				if (i < 0)
					i = 0;
				serversList.elementClicked(i, false, 0, 0);
				break;
			}
			case Keyboard.KEY_DOWN:
				int i = serversList.getSelectedSlot() + 1;
				if (i >= serversList.getSize())
					i = serversList.getSize() - 1;
				serversList.elementClicked(i, false, 0, 0);
				break;
			case Keyboard.KEY_RETURN:
				serversList.elementClicked(serversList.getSelectedSlot(), true, 0, 0);
				break;
			case Keyboard.KEY_NEXT:
				serversList.represented.scrollBy(representedScreen.getHeight() - 100);
				break;
			case Keyboard.KEY_PRIOR:
				serversList.represented.scrollBy(-representedScreen.getHeight() + 100);
				return;
		}

		super.keyTyped(typedChar, keyCode);
	}

	@Override
	public final void handleMouseInput() throws IOException
	{
		super.handleMouseInput();
		serversList.represented.handleMouseInput();
	}

	private class GuiServersList extends WrappedGuiSlot
	{
		private final MinecraftAccount account;
		int selectedSlot;

		GuiServersList(final GuiBannedServers prevGui, final MinecraftAccount acc)
		{
			super(mc, prevGui.representedScreen.getWidth(), prevGui.representedScreen.getHeight(), 40, prevGui.representedScreen.getHeight() - 40, 15);
			account = acc;
		}

		@Override
		public final boolean isSelected(final int id)
		{
			return selectedSlot == id;
		}

		final int getSelectedSlot()
		{
			if (selectedSlot > account.getBannedServers().size())
				selectedSlot = -1;
			return selectedSlot;
		}

		@Override
		public final int getSize()
		{
			return account.getBannedServers().size();
		}

		@Override
		public final void elementClicked(final int id, final boolean doubleClick, final int var3, final int var4)
		{
			selectedSlot = id;
		}

		@Override
		public final void drawSlot(final int id, final int x, final int y, final int var4, final int mouseXIn, final int mouseYIn)
		{
			final String server = account.getBannedServers().get(id);

			Fonts.font40.drawCenteredString(server, represented.getWidth() / 2, y + 1, Color.RED.getRGB(), true);
		}

		@Override
		public void drawBackground()
		{
		}
	}

	private static class GuiAddBanned extends WrappedGuiScreen
	{
		private final GuiBannedServers prevGui;
		private final MinecraftAccount account;
		private IGuiTextField name;
		private String status;

		private GuiAddBanned(final GuiBannedServers prevGui, final MinecraftAccount acc)
		{
			this.prevGui = prevGui;
			account = acc;
		}

		@Override
		public final void initGui()
		{
			final int width = representedScreen.getWidth();
			final int height = representedScreen.getHeight();

			Keyboard.enableRepeatEvents(true);
			representedScreen.getButtonList().add(classProvider.createGuiButton(1, width / 2 - 100, height / 4 + 96, "Add " + (account.getAccountName() == null ? account.getName() : account.getAccountName()) + "'s banned server"));
			representedScreen.getButtonList().add(classProvider.createGuiButton(0, width / 2 - 100, height / 4 + 120, "Back"));

			name = classProvider.createGuiTextField(2, Fonts.font40, width / 2 - 100, 60, 200, 20);
			name.setFocused(true);
			name.setText("");
			name.setMaxStringLength(128);
		}

		@Override
		public final void drawScreen(final int mouseX, final int mouseY, final float partialTicks)
		{
			final int width = representedScreen.getWidth();
			final int height = representedScreen.getHeight();

			representedScreen.drawBackground(0);
			RenderUtils.drawRect(30, 30, width - 30, height - 30, Integer.MIN_VALUE);

			Fonts.font40.drawCenteredString("Add Banned Server", width / 2, 34, 0xffffff);
			if (status != null)
				Fonts.font40.drawCenteredString(status, width / 2, height / 4 + 84, 0xffffff);
			name.drawTextBox();

			if (name.getText().isEmpty() && !name.isFocused())
				Fonts.font40.drawCenteredString("\u00A77Add Server", width / 2 - 74, 66, 0xffffff);

			super.drawScreen(mouseX, mouseY, partialTicks);
		}

		@Override
		public final void actionPerformed(final IGuiButton button) throws IOException
		{
			switch (button.getId())
			{
				case 0:
					mc.displayGuiScreen(prevGui.representedScreen);
					break;
				case 1:
					final String server = name.getText();
					if (server.isEmpty())
					{
						status = "\u00A7cEnter a server address!";
						return;
					}

					if (account.getBannedServers().contains(server))
					{
						status = "\u00A7cServer already exists!";
						return;
					}

					account.getBannedServers().add(server);
					LiquidBounce.fileManager.saveConfig(LiquidBounce.fileManager.accountsConfig);

					prevGui.status = "\u00A7aAdded banned server \u00A7c" + server + " \u00A7aof " + (account.getAccountName() == null ? account.getName() : account.getAccountName()) + "\u00A7c.";
					mc.displayGuiScreen(prevGui.representedScreen);
					break;
			}
			super.actionPerformed(button);
		}

		@Override
		public final void keyTyped(final char typedChar, final int keyCode) throws IOException
		{
			if (keyCode == Keyboard.KEY_ESCAPE)
			{
				mc.displayGuiScreen(prevGui.representedScreen);
				return;
			}

			if (name.isFocused())
				name.textboxKeyTyped(typedChar, keyCode);
			super.keyTyped(typedChar, keyCode);
		}

		@Override
		public final void mouseClicked(final int mouseX, final int mouseY, final int mouseButton) throws IOException
		{
			name.mouseClicked(mouseX, mouseY, mouseButton);
			super.mouseClicked(mouseX, mouseY, mouseButton);
		}

		@Override
		public final void updateScreen()
		{
			name.updateCursorCounter();
			super.updateScreen();
		}

		@Override
		public final void onGuiClosed()
		{
			Keyboard.enableRepeatEvents(false);
			super.onGuiClosed();
		}
	}
}

File: shared/main/java/net/ccbluex/liquidbounce/ui/client/altmanager/sub/GuiBannedServers.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.ui.client.altmanager.sub

import net.ccbluex.liquidbounce.LiquidBounce
import net.ccbluex.liquidbounce.api.minecraft.client.gui.IGuiButton
import net.ccbluex.liquidbounce.api.minecraft.client.gui.IGuiTextField
import net.ccbluex.liquidbounce.api.util.WrappedGuiScreen
import net.ccbluex.liquidbounce.api.util.WrappedGuiSlot
import net.ccbluex.liquidbounce.file.FileManager.Companion.saveConfig
import net.ccbluex.liquidbounce.ui.client.altmanager.GuiAltManager
import net.ccbluex.liquidbounce.ui.font.Fonts
import net.ccbluex.liquidbounce.utils.login.MinecraftAccount
import net.ccbluex.liquidbounce.utils.render.RenderUtils.drawRect
import org.lwjgl.input.Keyboard
import java.awt.Color
import java.io.IOException

class GuiBannedServers(private val prevGui: GuiAltManager, private val account: MinecraftAccount) : WrappedGuiScreen()
{
	private lateinit var serversList: GuiServersList

	var status = "\u00A77Idle..."

	override fun initGui()
	{
		val width = representedScreen.width
		val height = representedScreen.height

		serversList = GuiServersList(this, account)
		serversList.represented.registerScrollButtons(7, 8)

		representedScreen.buttonList.add(classProvider.createGuiButton(1, width - 80, 46, 70, 20, "Add"))
		representedScreen.buttonList.add(classProvider.createGuiButton(2, width - 80, 70, 70, 20, "Remove"))
		representedScreen.buttonList.add(classProvider.createGuiButton(0, width - 80, height - 65, 70, 20, "Back"))
	}

	override fun drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float)
	{
		val width = representedScreen.width

		representedScreen.drawBackground(0)
		serversList.represented.drawScreen(mouseX, mouseY, partialTicks)

		val middleScreen = width / 2f

		Fonts.font40.drawCenteredString("\u00A7cBanned servers\u00A78 of \u00A7a" + if (account.accountName == null) account.name else account.accountName, middleScreen, 6f, 0xffffff)
		Fonts.font35.drawCenteredString("\u00A7a" + (if (account.accountName == null) account.name else account.accountName) + "\u00A78 is \u00A7cbanned\u00A78 from \u00A7c" + serversList.getSize() + "\u00A7a servers.", middleScreen, 18f, 0xffffff)
		Fonts.font35.drawCenteredString(status, middleScreen, 32f, 0xffffff)

		super.drawScreen(mouseX, mouseY, partialTicks)
	}

	@Throws(IOException::class)
	override fun actionPerformed(button: IGuiButton)
	{
		when (button.id)
		{
			0 -> mc.displayGuiScreen(prevGui.representedScreen)
			1 -> mc.displayGuiScreen(GuiAddBanned(this, account).representedScreen)
			2 -> status = if (serversList.selectedSlot != -1 && serversList.selectedSlot < serversList.getSize())
			{
				account.bannedServers.remove(account.bannedServers[serversList.selectedSlot])
				saveConfig(LiquidBounce.fileManager.accountsConfig)
				"\u00A7aThe server has been removed."
			}
			else "\u00A7cSelect a server."
		}
		super.actionPerformed(button)
	}

	@Throws(IOException::class)
	override fun keyTyped(typedChar: Char, keyCode: Int)
	{
		when (keyCode)
		{
			Keyboard.KEY_ESCAPE ->
			{
				mc.displayGuiScreen(prevGui.representedScreen)
				return
			}

			Keyboard.KEY_UP ->
			{
				var i = serversList.selectedSlot - 1
				if (i < 0) i = 0
				serversList.elementClicked(i, false, 0, 0)
			}

			Keyboard.KEY_DOWN ->
			{
				var i = serversList.selectedSlot + 1
				if (i >= serversList.getSize()) i = serversList.getSize() - 1
				serversList.elementClicked(i, false, 0, 0)
			}

			Keyboard.KEY_RETURN -> serversList.elementClicked(serversList.selectedSlot, true, 0, 0)
			Keyboard.KEY_NEXT -> serversList.represented.scrollBy(representedScreen.height - 100)

			Keyboard.KEY_PRIOR ->
			{
				serversList.represented.scrollBy(-representedScreen.height + 100)
				return
			}
		}
		super.keyTyped(typedChar, keyCode)
	}

	@Throws(IOException::class)
	override fun handleMouseInput()
	{
		super.handleMouseInput()
		serversList.represented.handleMouseInput()
	}

	class GuiServersList internal constructor(prevGui: GuiBannedServers, private val account: MinecraftAccount) : WrappedGuiSlot(mc, prevGui.representedScreen.width, prevGui.representedScreen.height, 40, prevGui.representedScreen.height - 40, 15)
	{
		internal var selectedSlot = 0
			get()
			{
				if (field > account.bannedServers.size) selectedSlot = -1
				return field
			}

		override fun isSelected(id: Int): Boolean = selectedSlot == id

		override fun getSize(): Int = account.bannedServers.size

		override fun elementClicked(id: Int, doubleClick: Boolean, var3: Int, var4: Int)
		{
			selectedSlot = id
		}

		override fun drawSlot(id: Int, x: Int, y: Int, var4: Int, mouseXIn: Int, mouseYIn: Int)
		{
			val server = account.bannedServers[id]
			val middleScreen = represented.width / 2f

			Fonts.font40.drawCenteredString(server, middleScreen, y + 1f, Color.RED.rgb, true)
		}

		override fun drawBackground()
		{
		}
	}

	private class GuiAddBanned(private val prevGui: GuiBannedServers, private val account: MinecraftAccount) : WrappedGuiScreen()
	{
		private lateinit var name: IGuiTextField
		var status = "\u00A77Idle..."

		override fun initGui()
		{
			val width = representedScreen.width
			val height = representedScreen.height

			Keyboard.enableRepeatEvents(true)

			val buttonX = width / 2 - 100
			val quarterScreen = height / 4

			representedScreen.buttonList.add(classProvider.createGuiButton(1, buttonX, quarterScreen + 96, "Add " + (if (account.accountName == null) account.name else account.accountName) + "'s banned server"))
			representedScreen.buttonList.add(classProvider.createGuiButton(0, buttonX, quarterScreen + 120, "Back"))

			name = classProvider.createGuiTextField(2, Fonts.font40, buttonX, 60, 200, 20).apply {
				isFocused = true
				text = ""
				maxStringLength = 128
			}
		}

		override fun drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float)
		{
			val width = representedScreen.width
			val height = representedScreen.height

			val middleScreen = width / 2f
			val quarterScreen = height / 4f

			representedScreen.drawBackground(0)

			drawRect(30, 30, width - 30, height - 30, Int.MIN_VALUE)


			Fonts.font40.drawCenteredString("Add Banned Server", middleScreen, 34f, 0xffffff)
			Fonts.font40.drawCenteredString(status, middleScreen, quarterScreen + 84f, 0xffffff)

			name.drawTextBox()

			if (name.text.isEmpty() && !name.isFocused) Fonts.font40.drawCenteredString("\u00A77Add Server", middleScreen - 74f, 66f, 0xffffff)

			super.drawScreen(mouseX, mouseY, partialTicks)
		}

		@Throws(IOException::class)
		override fun actionPerformed(button: IGuiButton)
		{
			when (button.id)
			{
				0 -> mc.displayGuiScreen(prevGui.representedScreen)

				1 ->
				{
					val server = name.text

					if (server.isEmpty())
					{
						status = "\u00A7cEnter a server address!"
						return
					}

					if (account.bannedServers.contains(server))
					{
						status = "\u00A7cServer already exists!"
						return
					}

					account.bannedServers.add(server)

					saveConfig(LiquidBounce.fileManager.accountsConfig)

					prevGui.status = "\u00A7aAdded banned server \u00A7c" + server + " \u00A7aof " + (if (account.accountName == null) account.name else account.accountName) + "\u00A7c."

					mc.displayGuiScreen(prevGui.representedScreen)
				}
			}
			super.actionPerformed(button)
		}

		@Throws(IOException::class)
		override fun keyTyped(typedChar: Char, keyCode: Int)
		{
			if (keyCode == Keyboard.KEY_ESCAPE)
			{
				mc.displayGuiScreen(prevGui.representedScreen)
				return
			}
			if (name.isFocused) name.textboxKeyTyped(typedChar, keyCode)
			super.keyTyped(typedChar, keyCode)
		}

		@Throws(IOException::class)
		override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int)
		{
			name.mouseClicked(mouseX, mouseY, mouseButton)
			super.mouseClicked(mouseX, mouseY, mouseButton)
		}

		override fun updateScreen()
		{
			name.updateCursorCounter()
			super.updateScreen()
		}

		override fun onGuiClosed()
		{
			Keyboard.enableRepeatEvents(false)
			super.onGuiClosed()
		}
	}
}
